README.md
# Ternary_Contour
Plots a series of 3-dimensional data on a ternary diagram in the form of a contour map. Contours are based on 'spatial'-density of datapoints in the domain of the ternary plot. Multiple 'sets' of series of 3D data can be plotted on the same figure, each with different colours of their contour maps.

* requires download of Tern_Base function as well.

<!-- Options -->
## Options
 User can specify the following parameters when calling the function:
 - choose the colours of the contour colours, according to cmap available options (https://matplotlib.org/stable/tutorials/colors/colormaps.html).
 - choose the number of contours.
 - to plot contour lines.
 - to display values of the contour level values within the contour lines.
 - to plot the individual datapoints of each dataset.
 - to plot multiple datasets.
 - choose the end-members of the plot, which must match the content of the input dataset (systems are: [Si+Al, Mg, Fe], [Si+Al, Mg+Fe, O], or [S, Ni, Fe]).

User can also easily adjust plotting preferences (such as text size and stoichiometric line colour) by adjusting the variables defined within the first lines of the function, if so wishes.

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- Example output -->
## Example output

It should be noted that all data must be in the form of a list (of the different groups of data with different contour colours) of lists (of multiple datapoints for each contour) of lists (for each datapoint, for example [Si + Al, Mg, Fe]):

<br />
<div align="center">
 
 [ [ [A11, B11, C11], ... [A1n, B1n, .. C1n] ], ... [ [Am1, Bm1, Cm1], ... [Amn, Bmn, .. Cmn] ] ]
 
 </div>

- where n is the numer of datapoints in a given dataset (a particular colour of contours) and m is the number of datasets to plot individually (the number of individual contour plots of differing colours). Even if m = 1, data must be in the form of [ [A11, B11, .. C11], ... [A1n, B1n, C1n] ], where dataset[0] = [ [A11, B11, .. C11], ... [A1n, B1n, C1n] ] and not = [A11, B11, C11]. Note that for the ternary diagram, A is the top corner, B is the left corner, and C the right corner of the triangle (A = Si + Al, B = Fe, C = Mg for example)


<br />
<div align="center">

[![Product Name Screen Shot][product-screenshot]](https://example.com)
 
 </div>

This figures above were generated from the following user-specified options:
 - dataset = [[[A11, B11, C11], ...[A1n, B1n, C1n]], [[A21, B21, C21], ...[A2n, B2n, C2n]]] where A = Si+Al, B = Fe, and C = Mg at.%
     - red: data_1 (such as A11), blue: data_2 (such as A21)
 - NumLevels = 10
 - Colour = ['Blues', 'Reds']
 - type = 'silicate'
 - and differing data options for each sub-figure:
     - A) ContLines = 'n',  ContourValues = 'n',  ContColourFill = 'y',  DataPointDisp = 'y'
     - B) ContLines = 'y',  ContourValues = 'y',  ContColourFill = 'y',  DataPointDisp = 'n'
     - C) ContLines = 'n',  ContourValues = 'n',  ContColourFill = 'y',  DataPointDisp = 'n'

 
<p align="right">(<a href="#top">back to top</a>)</p>

<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community what it is. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/NewFeature`)
3. Commit your Changes (`git commit -m 'Add some NewFeature'`)
4. Push to the Branch (`git push origin feature/NewFeature`)
5. Open a Pull Request

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- CONTACT -->
## Contact

Dan Hallatt - daniel.hallatt@univ-lille.fr

Project Link: [https://github.com/DanHallatt/Ternary_Contour](https://github.com/DanHallatt/Ternary_Contour)

Associated Institute Link: https://umet.univ-lille.fr/MTP/index.php?lang=fr

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- Related repositories -->
## Related repositories
- plotting contours of compositions/metrics in 2D: https://github.com/DanHallatt/TwoDim_Contour
- making a video of dynamic ternary contours, as a window slides through the series of data: https://github.com/DanHallatt/Ternary_Contour_Vid
- making a video of dynamic contours compositions/metrics in 2D, as a window slides through the series of data: https://github.com/DanHallatt/TwoDim_Contour_Vid

<p align="right">(<a href="#top">back to top</a>)</p>


[product-screenshot]: Images/ExampleTernaryPlots.png


CHANGELOG.md
# Changelog

## v1.0.0

### Added or Changed
- Added this changelog :)

## v2.0.0

### Added or Changed
- Added multi-dataset capability.
- Added more user-defined options in both function calling and within function (plotting settings).
- Added transparent background of plots.


Tern_Base.py
# This function was originally written by Corentin Le Guillou, and later modified by Dan Hallatt, both at the Université de Lille.

import math
import matplotlib.pyplot as plt
import matplotlib.style as mpl_style
import matplotlib as mpl
import matplotlib.ticker as ticker
import numpy as np
import pandas as pd
from sigfig import round
from statistics import mean
import scipy.stats as st

def Tern_Base(type):
    """
    This is specifically the base style of the ternary plot.
    - defines axis positions, stoichiometric reference lines/locations.
    """
    
    StoichTextSize = 9.5
    StoichMarkerSize = 8
    StoichLineWidth = 1
    StoichLineColour = 'dimgrey'
    StoichTextColour = 'dimgrey'
    EndMemberTextSize = 13
    AxisTickSize = 5
    AxisLineWidth = 1
    
    fig, ax = plt.subplots()
    ax.plot([-0.5, -0.45, -0.4, -0.35, -0.3, -0.25, -0.2, -0.15, -0.1, -0.05, 0],  [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1], color = 'black', marker="_", markersize=AxisTickSize, linewidth=AxisLineWidth)
    ax.plot([0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5],  [1, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1, 0], color='black', marker="_",  markersize=AxisTickSize, linewidth=AxisLineWidth)
    ax.plot([-0.5, -0.4, -0.3, -0.2, -0.1, 0, 0.1, 0.2, 0.3, 0.4, 0.5], [0,0,0,0,0,0,0,0,0,0,0], color='black', marker="|",  markersize=AxisTickSize, linewidth=AxisLineWidth)
    
    if type == 'silicate':
        ax.plot([0.29, -0.29], [0.4, 0.4], linestyle=':', color=StoichLineColour, linewidth=StoichLineWidth, zorder=1)
        ax.annotate('Serp', xy=(0.3, 0.4), xytext = (0.32, 0.39), size=StoichTextSize, color=StoichTextColour)
        
        ax.plot([0.21, -0.21], [0.57, 0.57], linestyle=':', color=StoichLineColour, linewidth=StoichLineWidth, zorder=1)
        ax.annotate('Sap', xy=(0.21, 0.57), xytext = (0.23, 0.56), size=StoichTextSize, color=StoichTextColour)
        
        ax.plot([0.329, -0.329], [0.333, 0.333], linestyle='--', color=StoichLineColour, linewidth=StoichLineWidth, zorder=1)
        ax.annotate('Ol', xy=(0.333, 0.333), xytext = (0.35, 0.32), size=StoichTextSize, color=StoichTextColour)
        
        ax.plot([0.24, -0.24], [0.5, 0.5], linestyle='--', color=StoichLineColour, linewidth=StoichLineWidth, zorder=1)
        ax.annotate('Py', xy=(0.25, 0.50), xytext = (0.27, 0.49), size=StoichTextSize, color=StoichTextColour)
        
        ax.annotate('Si + Al', xy=(0.2, 1.), xytext = (-0.072, 1.03), size=EndMemberTextSize)
        ax.annotate('Fe', xy=(-0.55,0.0), xytext = (-0.57,-0.03), size=EndMemberTextSize)
        ax.annotate('Mg', xy=(0.55,0.0), xytext = (0.53, -0.03), size=EndMemberTextSize)

    if type == 'sulfide':
        ax.plot(-0.25, 0.5, 'ro', markersize=StoichMarkerSize)
        ax.annotate('Troilite', xy=(-0.25, 0.5), xytext = (-0.4, 0.5), size=StoichTextSize, color=StoichTextColour)
        
        ax.plot(0.0, 0.47, 'bo', markersize=StoichMarkerSize)
        ax.annotate('Pentlandite', xy=(0.0, 0.47), xytext = (0.03,0.47), size=StoichTextSize, color=StoichTextColour)
        
        ax.plot(-0.17,0.67, 'go',  markersize=StoichMarkerSize)
        ax.annotate('Pyrite', xy=(-0.17,0.67), xytext = (-0.32,0.67), size=StoichTextSize, color=StoichTextColour)
                        
        ax.plot([-0.25, -0.22], [0.5, 0.56], color='purple', linewidth=StoichLineWidth)
        ax.annotate('Pyrrhotite', xy=(-0.235, 0.55), xytext = (-0.42,0.55), size=StoichTextSize, color=StoichTextColour)
            
        ax.annotate('S', xy=(0., 1.0), xytext = (-0.05, 1.03), size=EndMemberTextSize)
        ax.annotate('Fe', xy=(-0.55,0.0), xytext = (-0.56,-0.03), size=EndMemberTextSize)
        ax.annotate('Ni', xy=(0.55,0.0), xytext = (0.52, -0.03), size=EndMemberTextSize)
        
    if type == 'silicateHydration':
        ax.plot(-0.21, 0.142857, marker="3", color= StoichLineColour, markersize=StoichMarkerSize, alpha=0.5)
        ax.annotate('Serp', xy=(-0.2, 0.14), xytext = (-0.235,0.05), size=StoichTextSize, color=StoichTextColour)
        
        ax.annotate('Si + Al', xy=(0.01, 1.0), xytext = (-0.06, 1.05), size=EndMemberTextSize)
        ax.annotate('O', xy=(-0.55,0.), xytext = (-0.57,-0.03), size=EndMemberTextSize)
        ax.annotate('Mg + Fe', xy=(0.55,0.0), xytext = (0.53, -0.03), size=EndMemberTextSize)

    return fig, ax


Ternary_Contour.py
# This function was originally written by Dan Hallatt, in-part using the ternary plotter made by Corentin Le Guillou. Both at the Université de Lille.

import numpy as np
import pandas as pd
import math
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.cm as cm
from matplotlib.ticker import (MultipleLocator, AutoMinorLocator)
import matplotlib.ticker as ticker
import matplotlib.tri as tri
import scipy.stats as st
from scipy.interpolate import griddata

from Tern_Base import Tern_Base

def Ternary_Contour(dataset, type, Colour, NumLevels, ContourValues, ContLines, ContColourFill, DataPointDisp, FigureSavePath, FileName):
    """ Uses matplotlib to plot a ternary diagram
    ** dataset : in the form of a list of list [ [ [A11, B11, C11], ... [A1n, B1n, .. C1n] ], [ [A21, B21, C21], ... [A2n, B2n, .. C2n] ] , ... [ [Am1, Bm1, Cm1], ... [Amn, Bmn, .. Cmn] ] ], where n is the numer of datapoints in a given dataset and m is the number of datasets to plot individually. Even if m = 1, data must be in form of [ [A11, B11, .. C11], ... [A1n, B1n, C1n] ], where dataset[0] = [ [A11, B11, .. C11], ... [A1n, B1n, C1n] ] and not = [A11, B11, C11]. Note that for the ternary diagram, A is the top corner, B is the left corner, and C the right corner of the triangle (A = Si + Al, B = Fe, C = Mg for example)
        
        !!! NOTE !!! : Data cannot contain 'NaN' values. Must be cleaned prior to input in this function (or this function modified to clean data by removing them).
        
    ** type : Either 'silicate' [Si+Al, Mg, Fe], 'silicateHydration' [Si+Al, Mg+Fe, O], or 'sulfide' [S, Ni, Fe].
    ** Colour : In quotations a list of the colour of the contour plots in form of ['Blues', 'Reds', ect..]. Number of colours defined must = m. Options available according to 'cmap' of matplotlib (https://matplotlib.org/stable/tutorials/colors/colormaps.html). Still required to be defined even if ContColourFill == 'n'.
    ** NumLevels: number of contour levels.
    ** ContourValues : Either 'y' (yes) or 'n' (no), to plot the values defining each contour level (value is density value). Requires ContLines='y'.
    ** ContLines : Either 'y' (yes) or 'n' (no), plot or not to plot the lines seperating contour levels.
    ** ContColourFill : Either 'y' (yes) or 'n' (no), to fill contour values with the colours defined in 'Colour' list.
    ** DataPointDisp : Either 'y' (yes) or 'n' (no), to plot the individual datapoints in the ternary diagram.
    ** FigureSavePath : Path to folder location where figures should be saved. Must be in single quotations, example : '/Volumes/Samsung_T5/Experiment categories/Laser/Figures/'
    ** FileName : General name of files to be saved. Must be in single quotations, example : 'TEST_DataSet01'
    ** ContourValues : Either 'y' (yes) or 'n' (no), plot or not to plot the values defining each contour level (value is density value). Requires ContLines='y'.
    """
    
    ContourLabelTextSize = 0.5
    ContourLineThickness = 0.4
    ContourLineStyle = '-'
    DataPointSize = 0.5
    
    # MAKING TERNARY DIAGRAM OF CONTOURS:
    fig, ax = Tern_Base(type)
    for k in range(0,len(dataset)):
        tri_y=[]
        tri_x=[]
        Data_Indiv = dataset[k]
        # Converting the 3D ternary data of [A, B, C] into 2D coordinates of [x,y] which correspond to the area within the ternary bounds. New data is thus tri_y and tri_x.
        for i in range (len(Data_Indiv)): # Number of ternary coordinates (Si, Fe, Mg trio's).
            tri_y.append(Data_Indiv[i][0]/(Data_Indiv[i][0]+Data_Indiv[i][1]+Data_Indiv[i][2]))
            tri_x.append(np.array(-0.5)*(np.array(1.)-tri_y[i])+(Data_Indiv[i][2]/(Data_Indiv[i][0]+Data_Indiv[i][1]+Data_Indiv[i][2])))
        y_Ternary = tri_y
        x_Ternary = tri_x
        xmin_Ternary, xmax_Ternary = -0.6, 0.6
        ymin_Ternary, ymax_Ternary = -0.1, 1.1

        # Peform the kernel density estimate
        xx_Ternary, yy_Ternary = np.mgrid[xmin_Ternary:xmax_Ternary:1000j, ymin_Ternary:ymax_Ternary:1000j]
        positions_Ternary = np.vstack([xx_Ternary.ravel(), yy_Ternary.ravel()])
        values_Ternary = np.vstack([x_Ternary, y_Ternary])
        kernel_Ternary = st.gaussian_kde(values_Ternary)
        f_Ternary = np.reshape(kernel_Ternary(positions_Ternary).T, xx_Ternary.shape)
        if ContColourFill == 'y':
            ContColour_Ternary = plt.contourf(xx_Ternary, yy_Ternary, f_Ternary, NumLevels, cmap=Colour[k], locator = ticker.MaxNLocator(prune = 'lower'), zorder=5)
        if ContLines == 'n' and ContourValues == 'y':
            ContourLineThickness = 0
            cset_Ternary = plt.contour(xx_Ternary, yy_Ternary, f_Ternary, NumLevels, colors='k', alpha=1, linewidths = ContourLineThickness, linestyles = ContourLineStyle, locator = ticker.MaxNLocator(prune = 'lower'), zorder=10) # Drawing contour lines.
            if ContourValues == 'y':
                ax.clabel(cset_Ternary, inline=1, fontsize=ContourLabelTextSize, zorder=15) # Labelling contour levels within the contour lines.
        if ContLines == 'y':
            cset_Ternary = plt.contour(xx_Ternary, yy_Ternary, f_Ternary, NumLevels, colors='k', alpha=1, linewidths = ContourLineThickness, linestyles = ContourLineStyle, locator = ticker.MaxNLocator(prune = 'lower'), zorder=10) # Drawing contour lines.
            if ContourValues == 'y':
                ax.clabel(cset_Ternary, inline=1, fontsize=ContourLabelTextSize, zorder=15) # Labelling contour levels within the contour lines.
        if DataPointDisp=='y':
            ax.scatter(tri_x, tri_y, color='black', alpha=1, s=DataPointSize, zorder=5)
            
    ax.axis('off')
    fig.savefig(FigureSavePath + FileName +  '_Ternary_Contour_' + type + 'Axis.pdf')

